include/pybind11/numpy.h

```diff
@@ -9,6 +9,12 @@
 
 #pragma once
 
+// Unless this is the first header PYBIND11_INTERNAL_NUMPY_1_ONLY_DETECTED is defined
```

Comment:
No, got it wrong again. This way it needs to be outside the include guard and then I don't need to change common.h, but the other way the numpy.h include cannot be the first.

Reply:
I undid it again for now.  Maybe I am missing something?  But I see the below as an option (untested), to avoid:
```
#include "pybind11/eigen/matrix.h"
#define PYBIND11_NUMPY_1_ONLY
#include "pybind11/numpy.h"
```
to give an error.  And I am not sure it is worth the trouble: 
```
// Use PYBIND11_NUMPY_H_ as include guard defined to `defined(PYBIND11_NUMPY_1_ONLY)`
// to ensure that we check that `PYBIND11_NUMPY_1_ONLY` is used consistently.
#ifdef PYBIND11_NUMPY_H_
#  if PYBIND11_NUMPY_H_ != defined(PYBIND11_NUMPY_1_ONLY)
#    error "PYBIND11_NUMPY_1_ONLY not defined before the first numpy.h include"
#  endif

#else  // PYBIND11_NUMPY_H_

#define PYBIND11_NUMPY_H_ defined(PYBIND11_NUMPY_1_ONLY)
#ifdef PYBIND11_NUMPY_1_ONLY
#    define  PYBIND11_INTERNAL_NUMPY_1_ONLY_DETECTED 1
#else
#    define  PYBIND11_INTERNAL_NUMPY_1_ONLY_DETECTED 0
#endif

#endif  // PYBIND11_NUMPY_H_

```

Reply:
Caveat: I didn't look to follow what you tried already.

> Maybe I am missing something?

Sorry I forgot to explain: you can take it as guaranteed that pybind11/detail/common.h is included before any other pybind11 include. (If not, that'd be a serious bug we'd want to fix.)

Please let me know if that doesn't resolve the situation and I'll look carefully.

Reply:
Maybe, this is the solution?
```
#ifndef PYBIND11_NUMPY_1_ONLY
#    define PYBIND11_NUMPY_1_ONLY "PYBIND11_NUMPY_1_ONLY not set before first include"
#    define PYBIND11_NUMPY_1_ONLY_INTERNAL 0
#else
#    define PYBIND11_NUMPY_1_ONLY_INTERNAL 1
#endif
```
Won't prevent things from compiling, but you will get a redefinition warning.  (Or just use `PYBIND11_NUMPY_1_ONLY 0/1` there, which just means you would have to use `#define PYBIND11_NUMPY_1_ONLY 1` if you put it in multiple places.